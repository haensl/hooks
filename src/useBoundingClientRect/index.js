import { useState, useCallback } from 'react';
import { useDebounce, useIsomorphicLayoutEffect, useWindowSize } from '../';

const useBoundingClientRect = (ref, debounceMs = 25) => {
  const [boundingClientRect, setBoundingClientRect] = useState(null);
  const windowSize = useWindowSize(debounceMs);

  const onResize = useCallback(({ target }) => {
    if (!target) {
      return;
    }

    const clientRect = target.getBoundingClientRect();
    setBoundingClientRect(clientRect);
  }, []);

  const onResizeDebounced = useDebounce(onResize, debounceMs);

  useIsomorphicLayoutEffect(() => {
    const current = ref.current;

    if (!current) {
      return;
    }

    const resizeObserver = new ResizeObserver(onResizeDebounced);
    resizeObserver.observe(current);
    onResize({ target: current });
    return () => {
      resizeObserver.unobserve(current);
    };
  }, [ ref, onResize, onResizeDebounced]);

  useIsomorphicLayoutEffect(() => {
    const current = ref && ref.current;
    if (!current || typeof onResize !== 'function') {
      return;
    }

    onResize({
      target: current
    });
  }, [ref, onResize, windowSize]);

  return boundingClientRect;
};

export default useBoundingClientRect;
