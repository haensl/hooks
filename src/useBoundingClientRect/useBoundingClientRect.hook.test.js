import { act, renderHook } from '@testing-library/react';
import '@testing-library/jest-dom';

import useBoundingClientRect from './';

let currentResizeObserver;

class ResizeObserver {
  constructor(callback) {
    this.targets = [];
    this.callback = callback;
    this.observe = jest.fn()
      .mockImplementation(function(target) {
        this.targets.push(target);
      }).bind(this);

    this.unobserve = jest.fn()
      .mockImplementation(function(target) {
        this.targets = this.targets.filter((t) => t !== target);
      }).bind(this);

    this.trigger = jest.fn()
      .mockImplementation(function() {
        for (const target of this.targets) {
          this.callback({ target });
        }
      }).bind(this);
    currentResizeObserver = this;
  }
}

const mockRef = () => {
  const node = document.createElement('div');
  const getBoundingClientRect = jest.fn()
    .mockReturnValue({
      top: 0,
      right: 0,
      bottom: 0,
      left: 0,
      width: 0,
      height: 0
    });
  node.getBoundingClientRect = getBoundingClientRect;
  const ref = {
    current: {
      getDOMNode: () => node,
      getBoundingClientRect
    }
  };

  return ref;
};

describe('useBoundingClientRect', () => {
  let originalResizeObserver;

  beforeAll(() => {
    jest.useFakeTimers();
    originalResizeObserver = global.ResizeObserver;
    global.ResizeObserver = ResizeObserver;
  });

  afterAll(() => {
    global.ResizeObserver = originalResizeObserver;
    jest.useRealTimers();
    jest.resetAllMocks();
  });

  test('renders without crashing', () => {
    const ref = mockRef();
    const { result } = renderHook(() => useBoundingClientRect(ref));
    expect(result.error)
      .not
      .toBeDefined();
  });

  describe('when the container is resized', () => {
    it('propagates the size change', () => {
      const ref = mockRef();
      const { result } = renderHook(() => useBoundingClientRect(ref));

      act(() => {
        const node = ref.current.getDOMNode();
        node.getBoundingClientRect
          .mockReturnValue({
            top: 0,
            right: 200,
            bottom: 200,
            left: 0,
            width: 200,
            height: 200
          });
        node.dispatchEvent(new Event('resize'));
        currentResizeObserver.trigger();
        jest.runOnlyPendingTimers();
      });

      expect(result.current)
        .toEqual(
          expect.objectContaining({
            width: 200,
            height: 200
          })
        );
    });
  });
});
