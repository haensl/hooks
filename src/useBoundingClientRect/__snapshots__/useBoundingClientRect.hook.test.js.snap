// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useBoundingClientRect renders as expected 1`] = `
<div>
  <span>
    no client rect
  </span>
</div>
`;

exports[`useBoundingClientRect when the container is resized renders as expected 1`] = `
<TestComponent>
  <div>
    <span>
      ,
      ,
      0
      ,
      0
    </span>
  </div>
</TestComponent>
`;
